### PR TITLE
doc: TRD: start non-XIP flash info document

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

From our [breakout](https://world.tockos.org/tockworld7/notes/breakout-non-xip/) at TW7, one item that came up was creating a document to overview what Tock might look like on platforms that do not support executing code stored in flash. The hope was to 1) clarify the goals and constraints so everyone is on the same page, and 2) to have some public documentation about this goal that others who might be interested can find.

This is a start towards such a document. This reflects a very basic understanding that I took away from the breakout session; I am by no means an expert on what these platforms look like.

### Testing Strategy

n/a

### TODO or Help Wanted

Ideally others would contribute text to expand this document.

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

File: doc/reference/trd-non-xip.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+This document provides information relating to running Tock on platforms without
+execute-in-place flash storage.
+
+Introduction
+===============================
+
+Traditional embedded systems use execute-in-place (XIP) flash to both store
+program code and to enable executing that program code. However, newer hardware
+designs, particularly those in security-focused settings, do not include any XIP
+flash storage. These platforms, subsequently referred to as "non-xip", only
```
Comment (alevy) [inline doc/reference/trd-non-xip.md:26]:
```suggestion
designs, particularly those integrated into a system-on-chip alongside an application processor, do not include 
executable persistent storage. These platforms, subsequently referred to as "non-xip", can only
```

File: doc/reference/trd-non-xip.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+execute instructions stored in volatile RAM. All executable code must be stored
+in external nonvolatile storage, loaded into RAM, and then executed.
```
Comment (alevy) [inline doc/reference/trd-non-xip.md:28]:
```suggestion
execute instructions stored in volatile RAM. All executable code must be loaded
from external nonvolatile storage, loaded into RAM, and then executed.
```

Comment (bradjc) [reply doc/reference/trd-non-xip.md:28]: Is it not stored anywhere?

File: doc/reference/trd-non-xip.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+Limitations:
+
+- Non-xip platforms have less RAM that traditional platforms have XIP flash.
+  Further, this RAM on non-xip platforms must (at runtime) store both the
+  executable code and the conventional memory items (e.g., the stack and heap).
+  This results in significantly less room for program code than on traditional
+  platforms.
+- As flash storage is not trusted, any data retrieved from flash storage must be
+  verified before it is used.
```
Comment (alevy) [inline doc/reference/trd-non-xip.md:70]:
```suggestion
- As flash storage may not be trustworthy, code retrieved from flash storage may need to be
  verified before it is used.
```

File: doc/reference/trd-non-xip.md (hunk 2 above)
Comment (bradjc) [reply doc/reference/trd-non-xip.md:28]: This was very confusing to me during the discussion, the code has to exist _somewhere_, no? Or is part of the assumption that the machine code may actually be completely lost at some point?

File: doc/reference/trd-non-xip.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+Non-limitations:
+
+- Energy is not a bottleneck, and additional power draw from having to load
+  program code into RAM is acceptable.
+- Execution time has relaxed deadlines, and additional execution time from
+  having to load program code into RAM is acceptable.
+
```
Comment (alevy) [inline doc/reference/trd-non-xip.md:78]:
```suggestion
- Loading code into RAM allows more predictable layout of code and data and/or rewriting some instructions during load time, potentially enabling simpler relocation and shared libraries.
```

File: doc/reference/trd-non-xip.md (hunk 2 above)
Comment (alevy) [reply doc/reference/trd-non-xip.md:28]: it's both true that it needs to be stored somewhere and that it needs to be loaded from somewhere. Between the two, I just think that the more salient part is that Tock needs to load it from somewhere.

**Comment by alevy** [comment]
People who were there (and who's Github handle I know): @lawrenceesswood, @lschuermann.

Note that I don't think it's necessary to get all the details in this document down now, but an opportunity to add or correct some stuff before we merge this initial draft.

File: doc/reference/trd-non-xip.md (hunk 2 above)
Comment (bradjc) [reply doc/reference/trd-non-xip.md:28]: I think it is helpful to clarify that the assumption is the code is stored somewhere, and even if it is evicted from RAM it could be retrieved again. If that assumption is _not_ expected to be true I think that could very materially change potential Tock designs.

**Comment by bradjc** [comment]
@reynoldsbd

**Comment by lschuermann** [comment]
> Note that I don't think it's necessary to get all the details in this document down now, but an opportunity to add or correct some stuff before we merge this initial draft.

To clarify: do we want to have this merged in the current form (i.e., skeleton + some context + motivation + concerns) or do we want to add actual technical details in this first iteration?

**Comment by bradjc** [comment]
I would like to add more. I think once this PR is merged the TRD is unlikely to change.

File: doc/reference/trd-non-xip.md (hunk 2 above)
Comment (lschuermann) [reply doc/reference/trd-non-xip.md:28]: I think that we shouldn't unnecessarily constrain ourselves by saying that Tock always has access to this code, which to me is implied by "stored". I think we want to support use-cases where a kernel has ephemeral access to binaries, or they are "streamed" over some interface (e.g., downloaded over TCP into memory). I think "loaded" encapsulates this a little better.

File: doc/reference/trd-non-xip.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  executable code and the conventional memory items (e.g., the stack and heap).
+  This results in significantly less room for program code than on traditional
+  platforms.
+- As flash storage may not be trustworthy, code retrieved from flash storage may need to be
+  verified before it is used.
+
+Non-limitations:
+
+- Energy is not a bottleneck, and additional power draw from having to load
+  program code into RAM is acceptable.
```
Comment (ppannuto) [inline doc/reference/trd-non-xip.md:75]: This may be too strong of a statement. IIUC, low-power operation is of interest and a pending priority for cases where Tock is an always-on component for battery-powered devices, e.g. laptops, phones.

That said, the energy overhead of additional RAM use may not be a significant factor; though, I think there was also some discussion of chips with banked SRAM and the capacity to power off unused portions, so it may yet be significant.

File: doc/reference/trd-non-xip.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+- Execution time has relaxed deadlines, and additional execution time from
+  having to load program code into RAM is acceptable.
```
Comment (ppannuto) [inline doc/reference/trd-non-xip.md:77]: Does this need to be qualified on "For some applications"?

File: doc/reference/trd-non-xip.md (hunk 2 above)
Comment (bradjc) [reply doc/reference/trd-non-xip.md:28]: In that case, that seems like a significant variation that we need to consider explicitly (and why this document exists).

File: doc/reference/trd-non-xip.md (hunk 6 above)
Comment (bradjc) [reply doc/reference/trd-non-xip.md:77]: These non-xip platforms have to give up _something_ right? They can't possibly be _as_ low power and _as_ fast with 25%-50% of the code storage as a conventional MCU can they? I can't imagine we can develop something if we can't assume some tradeoffs in all non-xip cases.

File: doc/reference/trd-non-xip.md (hunk 2 above)
Comment (lschuermann) [reply doc/reference/trd-non-xip.md:28]: I'm not sure whether we talked about this _explicitly_ in the breakout session during TW7, but it felt implied. For instance, applications may be provided in storage that's "leased" from some larger chip where that memory is either not always available, or may be modified at any given time (for our purposes, volatile external memory). To support these use-cases it seems like it's vital for Tock's support to _not_ require unconditional access to such external storage.

But it'd be great if others can confirm whether or not that's a real concern. @lawrenceesswood @reynoldsbd

Comment (bradjc) [reply doc/reference/trd-non-xip.md:28]: I just want to replace the implication with something concretely written down one way or the other because I think my assumptions have been generally different from the others in that breakout.

File: doc/reference/trd-non-xip.md (hunk 6 above)
Comment (ppannuto) [reply doc/reference/trd-non-xip.md:77]: Yeah, I think what I was probing at here was that _some_ (and likely few) applications could be time sensitive and would need to perhaps be 'pinned' or similar, which is another design point to consider.

File: doc/reference/trd-non-xip.md (hunk 2 above)
Comment (hudson-ayers) [reply doc/reference/trd-non-xip.md:28]: At work I often work on non-XIP MCU platforms, and one common use-case during development is the ability to stream code directly into RAM via JTAG and then execute from RAM, without having to store that code in flash at any intermediate point. This has several advantages:
1. it is faster
2. it reduces flash write cycles, which can be useful for hardware that is running hourly CI runs
3. if you have less flash than RAM, this enables running debug builds that only fit in RAM but do not fit in flash. This can be highly useful if your application is flash constrained in practice but you occasionally want to be able to run true debug builds (i.e. unoptimized code)

Of course, Tock could deign not to support this mode of operation. 

It is also possible on these platforms for OTA updates to overwrite the flash that contains the currently running application (since that application has already been copied into RAM). This is another common use case where Tock assuming that the application running in RAM matches the application that exists in flash could be problematic.

File: doc/reference/trd-non-xip.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+compared to traditional platforms, and the properties that may be important on
+traditional platforms but are likely not a priority on non-xip platforms.
+
+Limitations:
+
+- Non-xip platforms have less RAM that traditional platforms have XIP flash.
+  Further, this RAM on non-xip platforms must (at runtime) store both the
+  executable code and the conventional memory items (e.g., the stack and heap).
+  This results in significantly less room for program code than on traditional
+  platforms.
```
Comment (hudson-ayers) [inline doc/reference/trd-non-xip.md:68]: This is not universally true -- for example, TI sells the popular AM263x / AM263Px line of chips which have 2MB and 3MB of RAM, respectively.

File: doc/reference/trd-non-xip.md (hunk 6 above)
Comment (hudson-ayers) [reply doc/reference/trd-non-xip.md:77]: On many of these platforms program code is only loaded into RAM after reset, so that is the only time there is an execution delay. Otherwise, these platforms often run faster than XIP platforms, as SRAM accesses are generally faster than NOR flash accesses.

The need to "page in" applications from flash at runtime is definitely not universal to all non-XIP MCU platforms -- just those with significantly less flash than SRAM.

File: doc/reference/trd-non-xip.md (hunk 5 above)
Comment (alevy) [reply doc/reference/trd-non-xip.md:75]: True, though also to add to the list of things that may not be an important consideration in this context (but yes in others) is the energy overhead of copying memory back and forth between external/persistent storage and RAM, or the energy overhead of swapping things out.

Perhaps

```suggestion
- Energy is not as significant a bottleneck as in other settings, and additional power draw from having to load
  program code into RAM is acceptable.
```

File: doc/reference/trd-non-xip.md (hunk 6 above)
Comment (alevy) [reply doc/reference/trd-non-xip.md:77]: Yes, though @hudson-ayers, two specific use cases that came for discussions are those in which there actually is not enough SRAM to concurrently have loaded all the processes that one might want. E.g., TPM2 is just too darn big to fit with more than one or two other processes on a reasonably sized platform.

In one use case, swapping was mentioned as an option, in another, just killing apps on eviction would be acceptable, but both cases involve loading/unloading from SRAM at runtime.

And in both of these cases, I think _some_ applications would prefer to be pinned.

Basically both sides seem to be true here.

File: doc/reference/trd-non-xip.md (hunk 7 above)
Comment (alevy) [reply doc/reference/trd-non-xip.md:68]:
```suggestion
- Non-xip platforms store (at runtime) both the
  executable code and the conventional memory items (e.g., data, stack and heap).
  This results in significantly less room for program code than on traditional
  platforms.
```

Comment (alevy) [reply doc/reference/trd-non-xip.md:68]: Can confirm that understanding of one of the real use cases in question has order of MBs of RAM (another has 100s of KB so, again, both are correct).

File: doc/reference/trd-non-xip.md (hunk 6 above)
Comment (alistair23) [reply doc/reference/trd-non-xip.md:77]: Dynamically loading parts of apps is definitely an important use case. Tock [already supports that](https://github.com/tock/tock/blob/master/kernel/src/platform/platform.rs#L262) :)

**Comment by bradjc** [comment]
I think what I'm learning from this discussion (and perhaps exactly why I think we need this document) is that "non-XIP" is too broad in practice to be meaningful. I don't think we can effectively design solutions for all non-XIP MCUs. This is no different than Tock not being designed for all "traditional" MCUs (eg 8 bit cores, multi cores, arm A, etc).

I think we need to start with narrowing down non-XIP to a subset for this document. My understanding, from the breakout, is that there is a need to support the small RAM, untrusted nonxip flash use case.

File: doc/reference/trd-non-xip.md (hunk 5 above)
Comment (ppannuto) [reply doc/reference/trd-non-xip.md:75]: I actually have a half-baked research project with some preliminary data from a thing a MS student started for a quarter if you want to run down that alley further...

As newer MCUs integrate more interesting non-volatile storage options (hello MRAM on the Apollo4), the energy tradeoff gets much tighter. To a first order, for the Apollo4 it's around 1.5~2m where the tipping point occurs where it's cheaper in energy to write a byte out to MRAM and read it back than to preserve it in SRAM — for processes that wake less often, swapping them out and disabling banks of SRAM could save energy, at least on paper :).

**Comment by bradjc** [comment]
I took the liberty of updating the doc. Happy to bikeshed the name at some point.

**Comment by alevy** [comment]
@LawrenceEsswood @reynoldsbd feedback from both of you would be really good here.

**Comment by ppannuto** [comment]
@ZhekaS it sounds like this may also be relevant to your use case — we welcome any feedback you may have here as well!

**Comment by ZhekaS** [comment]
> @ZhekaS it sounds like this may also be relevant to your use case — we welcome any feedback you may have here as well!

It is very relevant indeed. Rather than providing concrete feedback, I'd rather describe the use-case.
 We have an SOC with many different cores on it, while at least one of them is expected to operate under Tock OS. All the firmwares are stored on external SPI-FLASH, and are cryptographically signed. In the boot-time relevant firmware is loaded into the specific component's SRAM where it is being validated (assuming SRAM is secure, it prevents TOCTOU which would be possible if validated in the flash). Then the firmware is being executed from SRAM. 
I do not see a big deal with this flow (but I am surely not an expert in TockOS :) ). As long as the TockOS is linked for the correct *SRAM* location, it should be working out of the box. The only thing, is that the startup code does not need to copy the `.data` section as it's load address should be the same as execution address in SRAM. Not sure how much MPU/PMP settings need to be adjusted.

**Comment by ppannuto** [comment]
This was discussed on the core team call today. While we would still like more feedback from downstream users, this is also version 1 of a draft document, and should not linger as a PR forever.

I'm going to attempt to resolve all of the open discussion threads now, and then ping folks for a merge.

File: doc/reference/trd-non-xip.md (hunk 2 above)
Comment (ppannuto) [reply doc/reference/trd-non-xip.md]: f22a2900bbd0c831d964837543328bd4f650c50c

File: doc/reference/trd-non-xip.md (hunk 5 above)
Comment (ppannuto) [reply doc/reference/trd-non-xip.md]: [trd-non-xip 9c7147b19] trd-non-xip: energy nuance comment

File: doc/reference/trd-non-xip.md (hunk 6 above)
Comment (ppannuto) [reply doc/reference/trd-non-xip.md]: [trd-non-xip 6b19cbbc3] trd-non-xip: capture dynamic app load nuances

**Review by ppannuto — APPROVED** [review]

**Comment by ppannuto** [comment]
Done — this should be good to go. Will let it sit for a few hours in case anyone sees anything in my edits here they don't like.

**Comment by bradjc** [comment]
Merge this?

**Comment by lschuermann** [comment]
Huh, I thought that I removed the litex-sim-ci checks from being required.

**Comment by ppannuto** [comment]
Rebased to pick up CI config changes.
